### PR TITLE
Multipath infrastructure

### PR DESCRIPTION
**Pull request by aznair**
These are the changes to the python code I need for my multipath socket code (to come later).

PathCombinator:
- I added a list of interfaces (isd_ad, if_number) when building paths. This is so the socket can know exactly which physical links a packet will traverse. This information will eventually be used for advanced congestion control.

SCION Daemon:
- Core paths are now fetched whether shortcuts are available or not for completeness. All possible core paths are fetched instead of using whatever core segment is first in the list. Also, get_paths will always request data from the path server because in my scenarios it is called when the socket specifically wants data that is as fresh as possible.

Protocol numbers:
- I've used 150 for SDAMP (multipath with congestion control) and 151 for SUDP (single path, no congestion control, path failure recovery via probing) arbitrarily.
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40434451%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40434504%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40435921%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40436205%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40436258%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40436714%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40436916%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40437730%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40491175%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40491176%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40491185%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23issuecomment-143406890%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%205a9a6253491806e0272e8ab55f796259dca49b1d%20lib/packet/path.py%2030%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40435921%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%60if%20egress%60%20%3F%22%2C%20%22created_at%22%3A%20%222015-09-25T14%3A35%3A34Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Interface%20number%200%20isn%27t%20actually%20a%20physical%20link%20so%20I%27m%20excluding%20those%22%2C%20%22created_at%22%3A%20%222015-09-25T14%3A38%3A34Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%2C%20%7B%22body%22%3A%20%22yes%2C%20I%20know%2C%20but%20instead%20of%20%60if%20egress%20%21%3D%200%3A%60%20you%20should%20write%20%60if%20egress%3A%60%22%2C%20%22created_at%22%3A%20%222015-09-25T14%3A44%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222015-09-26T07%3A34%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/path.py%3AL817-845%22%7D%2C%20%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23issuecomment-143406890%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22lgtm%22%2C%20%22created_at%22%3A%20%222015-09-26T07%3A35%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%205a9a6253491806e0272e8ab55f796259dca49b1d%20endhost/sciond.py%2020%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40434451%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22could%20be%20%60if%20self.up_segments...%60%22%2C%20%22created_at%22%3A%20%222015-09-25T14%3A23%3A00Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22just%20noticed%20that%20it%20is%20an%20older%20code%20%3B-%29%22%2C%20%22created_at%22%3A%20%222015-09-25T14%3A23%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222015-09-25T14%3A51%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/sciond.py%3AL220-255%22%7D%2C%20%22Pull%205a9a6253491806e0272e8ab55f796259dca49b1d%20lib/packet/path.py%2097%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40436714%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22not%20sure%20why%20you%20want%20to%20append%20this%20and%20the%20next%20interface%22%2C%20%22created_at%22%3A%20%222015-09-25T14%3A42%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222015-09-26T07%3A35%3A37Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/path.py%3AL918-971%22%7D%2C%20%22Pull%205a9a6253491806e0272e8ab55f796259dca49b1d%20endhost/sciond.py%2082%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/387%23discussion_r40436205%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22this%20description%20is%20inconsistent%20with%20the%20code%22%2C%20%22created_at%22%3A%20%222015-09-25T14%3A38%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222015-09-26T07%3A34%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/sciond.py%3AL306-313%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/387#issuecomment-143406890'>General Comment</a></b>
- <a href='https://github.com/pszalach'></a> lgtm
- [x] <a href='#crh-comment-Pull 5a9a6253491806e0272e8ab55f796259dca49b1d endhost/sciond.py 20'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/387#discussion_r40434451'>File: endhost/sciond.py:L220-255</a></b>
- <a href='https://github.com/pszalach'></a> could be `if self.up_segments...`
- <a href='https://github.com/pszalach'></a> just noticed that it is an older code ;-)
- [x] <a href='#crh-comment-Pull 5a9a6253491806e0272e8ab55f796259dca49b1d lib/packet/path.py 30'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/387#discussion_r40435921'>File: lib/packet/path.py:L817-845</a></b>
- <a href='https://github.com/pszalach'></a> `if egress` ?
- <a href='https://github.com/aznair'></a> Interface number 0 isn't actually a physical link so I'm excluding those
- <a href='https://github.com/pszalach'></a> yes, I know, but instead of `if egress != 0:` you should write `if egress:`
- [x] <a href='#crh-comment-Pull 5a9a6253491806e0272e8ab55f796259dca49b1d endhost/sciond.py 82'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/387#discussion_r40436205'>File: endhost/sciond.py:L306-313</a></b>
- <a href='https://github.com/pszalach'></a> this description is inconsistent with the code
- [x] <a href='#crh-comment-Pull 5a9a6253491806e0272e8ab55f796259dca49b1d lib/packet/path.py 97'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/387#discussion_r40436714'>File: lib/packet/path.py:L918-971</a></b>
- <a href='https://github.com/pszalach'></a> not sure why you want to append this and the next interface

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/387?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/387?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/387'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: endhost/sciond.py [hunk 1]
```diff
@@ -220,43 +220,35 @@ def get_paths(self, dst_isd, dst_ad, requester=None):
...
         """
         full_paths = []
+        self._request_paths(PST.UP_DOWN, dst_isd, dst_ad, requester=requester)
         down_segments = self.down_segments(last_isd=dst_isd, last_ad=dst_ad)
-        # Fetch down-paths if necessary.
-        if not down_segments:
-            self._request_paths(PST.UP_DOWN, dst_isd, dst_ad,
-                                requester=requester)
-            down_segments = self.down_segments(last_isd=dst_isd, last_ad=dst_ad)
         if len(self.up_segments) and down_segments:
```
Comment (pszal) [inline endhost/sciond.py:225]: could be `if self.up_segments...`

Comment (pszal) [inline endhost/sciond.py:225]: just noticed that it is an older code ;-)

File: lib/packet/path.py [hunk 2]
```diff
@@ -815,8 +817,28 @@ def _build_core_path(cls, up_segment, core_segment, down_segment):
...
+        path = CorePath.from_values(up_iof, up_hofs, core_iof, core_hofs,
                                     down_iof, down_hofs)
+        up_core = up_segment.ads
+        if core_segment:
+            up_core = up_core + core_segment.ads
+        for block in up_core:
+            isd_ad = ISD_AD(block.pcbm.isd_id, block.pcbm.ad_id)
+            egress = block.pcbm.hof.egress_if
+            ingress = block.pcbm.hof.ingress_if
+            if egress != 0:
```
Comment (pszal) [inline lib/packet/path.py]: `if egress` ?

File: endhost/sciond.py [hunk 3]
```diff
@@ -314,7 +306,7 @@ def _api_handle_path_request(self, packet, sender):
         Path request:
           | \x00 (1B) | ISD (12bits) |  AD (20bits)  |
         Reply:
-          |path1_len(1B)|path1(path1_len*8B)|first_hop_IP(4B)|path2_len(1B)...
+          |p1_len(1B)|p1((p1_len*8)B)|fh_IP(4B)|fh_port(2B)|p1_mtu(4B)|p1_RTT(4B)|p2_len(1B)...
```
Comment (pszal) [inline endhost/sciond.py]: this description is inconsistent with the code

File: lib/packet/path.py (hunk 2 above)
Comment (aznair) [inline lib/packet/path.py]: Interface number 0 isn't actually a physical link so I'm excluding those

Comment (pszal) [inline lib/packet/path.py]: yes, I know, but instead of `if egress != 0:` you should write `if egress:`

File: endhost/sciond.py (hunk 1 above)
Comment (pszal) [inline endhost/sciond.py:225]: :+1:

File: endhost/sciond.py (hunk 3 above)
Comment (pszal) [inline endhost/sciond.py]: :+1:

File: lib/packet/path.py (hunk 2 above)
Comment (pszal) [inline lib/packet/path.py]: :+1:

File: lib/packet/path.py [hunk 4]
```diff
@@ -896,24 +918,53 @@ def _join_shortcuts(cls, up_segment, down_segment, point, peer=True):
...
+                path.interfaces.append((isd_ad, egress))
+            if i != up_index:
+                path.interfaces.append((isd_ad, ingress))
+        if peer:
+            up_pcbm = up_segment.ads[up_index].pcbm
+            up_isd_ad = ISD_AD(up_pcbm.isd_id, up_pcbm.ad_id)
+            down_pcbm = down_segment.ads[down_index].pcbm
+            down_isd_ad = ISD_AD(down_pcbm.isd_id, down_pcbm.ad_id)
+            path.interfaces.append((up_isd_ad, up_peering_hof.ingress_if))
```
Comment (pszal) [inline lib/packet/path.py:956]: :+1:

Comment (pszal) [inline lib/packet/path.py:956]: :+1:

**Pull request merged**